presto-main/src/main/java/io/prestosql/sql/planner/ScopeAware.java

```diff
@@ -0,0 +1,169 @@
+    {
+        return new ScopeAware<T>(analysis, scope, node);
+    }
+
+    public T getNode()
+    {
+        return node;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return hash(node);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ScopeAware<T> other = (ScopeAware<T>) o;
+        checkArgument(scope.isLocalScope(other.scope) || other.scope.isLocalScope(scope), "Expressions must be in the same local scope");
+
+        return equal(node, other.node);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ScopeAware(" + node + ")";
+    }
+
+    private boolean equal(Node left, Node right)
+    {
+        if (left instanceof Expression && right instanceof Expression) {
+            Expression leftExpression = (Expression) left;
+            Expression rightExpression = (Expression) right;
+            if (analysis.isColumnReference(leftExpression) && analysis.isColumnReference(rightExpression)) {
+                ResolvedField leftField = analysis.getResolvedField(leftExpression);
+                ResolvedField rightField = analysis.getResolvedField(rightExpression);
+
+                Scope leftScope = leftField.getScope();
+                Scope rightScope = rightField.getScope();
+
+                if (leftScope != rightScope) {
+                    if (leftScope.hasOuterParent(scope) && rightScope.hasOuterParent(scope)) {
+                        // They are local to subqueries, but come from different subqueries
+                        // The specific fields they reference don't matter as long as everything else is syntactically equal in the subquery
+                        // TODO: eventually, equality should be based on tracing references to tables/views/CTEs
+                        return leftExpression.equals(rightExpression);
+                    }
+                }
+
+                if (isLocalScope(leftScope)) { // pick any, both are equal
+                    // fields come from the same scope as the original expression, so compare by field id
+                    return leftField.getFieldId().equals(rightField.getFieldId());
+                }
+
+                // Otherwise, they are either:
+                // 1. local to the subquery
+                // 2. from a scope that is a parent to the original scope (we verify that the top-level expressions
+                //    being compared come from the same local scope)
+                //
+                // In the first case, the specific fields they reference don't matter as long as everything else is syntactically equal.
+                // In the second case, the fields must necessarily be the same if they are syntactically equal.
+                return leftExpression.equals(rightExpression);
```

Comment:
I would rewrite it as:

```
                // case 1. no equality when corresponding column references refer to fields from different level scopes
                if (leftScope.hasOuterParent(rightScope) || rightScope.hasOuterParent(leftScope)) {
                    return false;
                }

                // case 2. both column references were resolved within the original scope:
                // we cannot rely on syntactical comparison and must take into account local scope hierarchy
                // (nested local scopes defined by FROM and SELECT clauses)
                if (isLocalScope(leftScope)) {
                    return leftField.getFieldId().equals(rightField.getFieldId());
                }

                // case 3. both column references were resolved within the same outer scope
                // case 4. column references were resolved within the same or different subquery scopes
                // syntactical comparison is sufficient
                return leftExpression.equals(rightExpression);
```

Reply:
I think it's much simpler than that:

```java
if (isLocalScope(leftScope) || isLocalScope(rightScope)) {
    return leftField.getFieldId().equals(rightField.getFieldId());
}

return leftExpression.equals(rightExpression);
```

The "case 1" in your example is just a short-circuit, but not necessary for correctness.

Reply:
This is correct, but not in the most obvious way.
If we check that _either_ scope is local and then compare `fieldId`s, we get
- local-scope-hierarchy-aware comparison in case both fields were local
- `false` in case the other field was not local

With this assumption, the if-condition could be simply `isLocalScope(leftScope)`as well.


I think we could write it differently just for the sake of clarity.
And I can see two options here, depending of what the idea is behind choosing the comparison strategy (syntax vs by-fieldId)

- idea 1: apply syntactical comparison whenever possible and compare `fieldId`s only within local scope
  ==> the if-condition would be `isLocalScope(leftScope) && isLocalScope(rightScope)`

- idea 2: compare by `fieldId` whenever possible and apply syntactical comparison only for fields local to subqueries. Such strategy would be easy to justify like this:
"fieldId is how we normally identify fields. However, with subqueries we want to take advantage if they are identical and that's why we change strategy to syntactical comparison (only for fields local to subqueries)"
in this case the code would look like:
```
if (leftScope.hasOuterParent(scope) && rightScope.hasOuterParent(scope)) {
      return leftExpression.equals(rightExpression);
      }
return leftField.getFieldId().equals(rightField.getFieldId());
```

Reply:
Rename `leftScope` -> `leftFieldScope`, `rightScope` -> `rightFieldScope`.
This will help keep clear distinction between the "original" scope of the analysed Nodes and the resolved scopes of particular fields.

Reply:
Unfortunately, the latter approach doesn't work. The problem is the following. Given this scope structure:

```
FROM     <=== SubQuery 1
  ^
  |
  |
ORDER BY <=== SubQuery 2
```

Where `===>` are outer scope relations and `--->` are local scope relations, when comparing an expression in `SubQuery1` scope with the key scope as `ORDER BY` scope,`ORDER BY` scope is NOT an outer parent of `Subquery 1` scope in the scope tree, but `SubQuery 1` is a child of the composite "local" scope made up of `FROM` and `ORDER BY`. `hasOuterParent` would need check if any of the local scopes in the outer parent scope is a parent. To make this work, we'd need to change how the scope hierarchy is represented. I plan to do a cleanup of that later, but for now, I'll have to leave this code as is.